src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioWorkspaceDiagnosticAnalyzerProviderService.cs

```diff
@@ -76,61 +73,74 @@ internal static IAnalyzerAssemblyLoader GetLoader()
+                    var extension_HeaderType = extension_Header.GetType();
+                    var extension_HeaderType_LocalizedNameProperty = extension_HeaderType.GetRuntimeProperty("LocalizedName");
+                    var name = extension_HeaderType_LocalizedNameProperty.GetValue(extension_Header) as string;
+
+                    var assemblies = ImmutableArray.CreateBuilder<string>();
 
-                    var extensionType_GetContentMethod = extensionType.GetRuntimeMethod("GetContentLocation", new Type[] { parameterType });
-                    var assembly = extensionType_GetContentMethod?.Invoke(extension, new object[] { content }) as string;
-                    if (assembly == null)
+                    // var extension_Content = extension.Content;
+                    var extensionType_ContentProperty = extensionType.GetRuntimeProperty("Content");
+                    var extension_Content = extensionType_ContentProperty.GetValue(extension) as IEnumerable<object>;
+
+                    foreach (var content in extension_Content)
                     {
-                        continue;
+                        if (!ShouldInclude(content))
+                        {
+                            continue;
+                        }
+
+                        var extensionType_GetContentMethod = extensionType.GetRuntimeMethod("GetContentLocation", new Type[] { parameterType });
+                        var assembly = extensionType_GetContentMethod?.Invoke(extension, new object[] { content }) as string;
+                        if (assembly == null)
+                        {
+                            continue;
+                        }
+
+                        assemblies.Add(assembly);
                     }
 
-                    assemblies.Add(assembly);
+                    builder.Add(new HostDiagnosticAnalyzerPackage(name, assemblies.ToImmutable()));
                 }
 
-                builder.Add(new HostDiagnosticAnalyzerPackage(name, assemblies.ToImmutable()));
-            }
-
-            var packages = builder.ToImmutable();
+                var packages = builder.ToImmutable();
 
-            EnsureMandatoryAnalyzers(packages);
+                EnsureMandatoryAnalyzers(packages);
 
-            // make sure enabled extensions are alive in memory
-            // so that we can debug it through if mandatory analyzers are missing
-            GC.KeepAlive(enabledExtensions);
+                // make sure enabled extensions are alive in memory
+                // so that we can debug it through if mandatory analyzers are missing
+                GC.KeepAlive(enabledExtensions);
 
-            return packages;
+                return packages;
+            }
+            catch (InvalidOperationException)
+            {
+                // this can be called from any thread, and extension manager could be disposed in the middle of us using it since
```

Comment:
Are we confident this is the only exception that can be thrown? What if they dispose in the middle of the call? Can we do something a bit more reliable here?

Reply:
such as?

I asked them whether there is other way to check whether the object is disposed. but I looked their code and seems there is no API to call to check.

Reply:
I would assume we'd do something here where we either wait for this to complete during shutdown (so it's never racing against their disposal), or we cancel it or something else. This seems really icky.

Reply:
> I asked them whether there is other way to check whether the object is disposed. but I looked their code and seems there is no API to call to check.

Correctly: there'd be a race between when you asked if it was disposed and a potential disposal. :smile: